File: swastricare-mobile-swift/Services/AnalyticsService.swift
//
//  AnalyticsService.swift
//  swastricare-mobile-swift
//
//  Firebase Analytics wrapper for screen views and custom events.
//

import Foundation
import FirebaseAnalytics

/// Wraps Firebase Analytics for consistent event logging across the app.
enum AnalyticsService {
    
    // MARK: - Screen tracking
    
    /// Log a screen view (e.g. when user opens a tab or view).
    static func logScreen(name: String, screenClass: String? = nil) {
        Analytics.logEvent(AnalyticsEventScreenView, parameters: [
            AnalyticsParameterScreenName: name,
            AnalyticsParameterScreenClass: screenClass ?? name
        ])
    }
    
    // MARK: - Custom events
    
    /// Log a custom event with optional parameters.
    static func logEvent(_ name: String, parameters: [String: Any]? = nil) {
        if let parameters = parameters {
            Analytics.logEvent(name, parameters: parameters)
        } else {
            Analytics.logEvent(name, parameters: nil)
        }
    }
    
    /// Log when user completes onboarding.
    static func logOnboardingComplete() {
        logEvent("onboarding_complete")
    }
    
    /// Log when user starts a workout/run.
    static func logWorkoutStart(activityType: String) {
        logEvent("workout_start", parameters: ["activity_type": activityType])
    }
    
    /// Log when user completes a workout.
    static func logWorkoutComplete(durationSeconds: Int, activityType: String) {
        logEvent("workout_complete", parameters: [
            "duration_seconds": durationSeconds,
            "activity_type": activityType
        ])
    }
    
    /// Log hydration log.
    static func logHydrationLogged(amountMl: Int) {
        logEvent("hydration_logged", parameters: ["amount_ml": amountMl])
    }
    
    /// Log medication reminder interaction.
    static func logMedicationReminder(action: String, medicationId: String? = nil) {
        var params: [String: Any] = ["action": action]
        if let id = medicationId { params["medication_id"] = id }
        logEvent("medication_reminder", parameters: params)
    }
    
    /// Set user ID for analytics (e.g. after login). Pass nil to clear.
    static func setUserId(_ userId: String?) {
        Analytics.setUserID(userId)
    }
    
    /// Set a user property for segmentation.
    static func setUserProperty(_ value: String?, forName name: String) {
        Analytics.setUserProperty(value, forName: name)
    }
}

File: android/app/src/main/kotlin/com/swastricare/health/data/services/AnalyticsService.kt
package com.swastricare.health.data.services

import android.os.Bundle
import android.util.Log
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.analytics.ktx.analytics
import com.google.firebase.ktx.Firebase

/**
 * Firebase Analytics wrapper service.
 * Matches iOS AnalyticsService pattern.
 * Gracefully handles cases where Firebase is not configured (missing google-services.json).
 */
class AnalyticsService {

    private val analytics: FirebaseAnalytics? = try {
        Firebase.analytics
    } catch (e: Exception) {
        Log.w(TAG, "Firebase Analytics not available: ${e.message}")
        null
    }

    // ─────────────────────────────────────
    // MARK: - Screen Tracking
    // ─────────────────────────────────────

    fun logScreenView(screenName: String) {
        val bundle = Bundle().apply {
            putString(FirebaseAnalytics.Param.SCREEN_NAME, screenName)
            putString(FirebaseAnalytics.Param.SCREEN_CLASS, screenName)
        }
        analytics?.logEvent(FirebaseAnalytics.Event.SCREEN_VIEW, bundle)
    }

    // ─────────────────────────────────────
    // MARK: - Generic Event Logging
    // ─────────────────────────────────────

    fun logEvent(eventName: String, params: Map<String, Any> = emptyMap()) {
        val bundle = Bundle().apply {
            params.forEach { (key, value) ->
                when (value) {
                    is String -> putString(key, value)
                    is Int -> putInt(key, value)
                    is Long -> putLong(key, value)
                    is Double -> putDouble(key, value)
                    is Float -> putFloat(key, value)
                    is Boolean -> putBoolean(key, value)
                    else -> putString(key, value.toString())
                }
            }
        }
        analytics?.logEvent(eventName, bundle)
    }

    // ─────────────────────────────────────
    // MARK: - Workout Events
    // ─────────────────────────────────────

    fun logWorkoutStart(activityType: String) {
        logEvent("workout_start", mapOf("activity_type" to activityType))
    }

    fun logWorkoutComplete(activityType: String, duration: Long, distance: Double) {
        logEvent("workout_complete", mapOf(
            "activity_type" to activityType,
            "duration_seconds" to duration,
            "distance_km" to distance
        ))
    }

    // ─────────────────────────────────────
    // MARK: - Hydration Events
    // ─────────────────────────────────────

    fun logHydrationLogged(amount: Int, drinkType: String) {
        logEvent("hydration_logged", mapOf(
            "amount_ml" to amount,
            "drink_type" to drinkType
        ))
    }

    // ─────────────────────────────────────
    // MARK: - Medication Events
    // ─────────────────────────────────────

    fun logMedicationTaken(medicationName: String) {
        logEvent("medication_taken", mapOf("medication_name" to medicationName))
    }

    fun logMedicationSkipped(medicationName: String, reason: String?) {
        logEvent("medication_skipped", mapOf(
            "medication_name" to medicationName,
            "reason" to (reason ?: "none")
        ))
    }

    // ─────────────────────────────────────
    // MARK: - AI Events
    // ─────────────────────────────────────

    fun logAIMessageSent(mode: String) {
        logEvent("ai_message_sent", mapOf("mode" to mode))
    }

    // ─────────────────────────────────────
    // MARK: - Vault Events
    // ─────────────────────────────────────

    fun logVaultUpload(category: String) {
        logEvent("vault_upload", mapOf("category" to category))
    }

    // ─────────────────────────────────────
    // MARK: - User Identity
    // ─────────────────────────────────────

    fun setUserId(userId: String) {
        analytics?.setUserId(userId)
    }

    fun setUserProperties(properties: Map<String, String>) {
        properties.forEach { (key, value) ->
            analytics?.setUserProperty(key, value)
        }
    }

    companion object {
        private const val TAG = "AnalyticsService"
    }
}
